crates/bevy_pbr/src/ssao/mod.rs

```diff
@@ -0,0 +1,958 @@
+///
+/// This darkens creases, e.g. on staircases, and gives nice contact shadows
+/// where objects meet, giving entities a more "grounded" feel.
+///
+/// # Usage Notes
+///
+/// Requires that you add [`ScreenSpaceAmbientOcclusionPlugin`] to your app,
+/// and add the [`DepthPrepass`] and [`NormalPrepass`] components to your camera.
+///
+/// It strongly recommended that you use SSAO in conjunction with
+/// TAA ([`bevy_core_pipeline::experimental::taa::TemporalAntiAliasSettings`]).
+/// Doing so greatly reduces SSAO noise.
+#[derive(Component, ExtractComponent, Reflect, PartialEq, Eq, Hash, Clone, Default)]
+pub enum ScreenSpaceAmbientOcclusionSettings {
+    Low,
+    Medium,
+    #[default]
+    High,
+    Ultra,
+    Custom {
+        /// Higher slice count means less noise, but worse performance.
+        slice_count: u32,
+        /// Samples per slice side is also tweakable, but recommended to be left at 2 or 3.
+        samples_per_slice_side: u32,
+    },
+}
+
+impl ScreenSpaceAmbientOcclusionSettings {
+    fn sample_counts(&self) -> (u32, u32) {
+        match self {
+            ScreenSpaceAmbientOcclusionSettings::Low => (1, 2), // 4 spp (1 * (2 * 2)), plus optional temporal samples
+            ScreenSpaceAmbientOcclusionSettings::Medium => (2, 2), // 8 spp (2 * (2 * 2)), plus optional temporal samples
+            ScreenSpaceAmbientOcclusionSettings::High => (3, 3), // 18 spp (3 * (3 * 2)), plus optional temporal samples
+            ScreenSpaceAmbientOcclusionSettings::Ultra => (9, 3), // 54 spp (9 * (3 * 2)), plus optional temporal samples
+            ScreenSpaceAmbientOcclusionSettings::Custom {
+                slice_count: slices,
+                samples_per_slice_side,
+            } => (*slices, *samples_per_slice_side),
+        }
+    }
+}
+
+#[derive(Default)]
+struct SSAONode {}
+
+impl ViewNode for SSAONode {
+    type ViewQuery = (
+        &'static ExtractedCamera,
+        &'static SSAOPipelineId,
+        &'static SSAOBindGroups,
+        &'static ViewUniformOffset,
+    );
+
+    fn run(
+        &self,
+        _graph: &mut RenderGraphContext,
+        render_context: &mut RenderContext,
+        (camera, pipeline_id, bind_groups, view_uniform_offset): QueryItem<Self::ViewQuery>,
+        world: &World,
+    ) -> Result<(), NodeRunError> {
+        let pipelines = world.resource::<SSAOPipelines>();
+        let pipeline_cache = world.resource::<PipelineCache>();
+        let (
+            Some(preprocess_depth_pipeline),
+            Some(spatial_denoise_pipeline),
+        ) = (
+            pipeline_cache.get_compute_pipeline(pipelines.preprocess_depth_pipeline),
+            pipeline_cache.get_compute_pipeline(pipelines.spatial_denoise_pipeline),
+        ) else {
+            return Ok(());
+        };
+        let Some(gtao_pipeline) = pipeline_cache.get_compute_pipeline(pipeline_id.0) else { return Ok(()) };
+        let Some(camera_size) = camera.physical_viewport_size else { return Ok(()) };
```

Comment:
Nit: could those two `let Some` be added to the two previous ones? so that it reads:

```rust
        let (
            Some(preprocess_depth_pipeline),
            Some(spatial_denoise_pipeline),
            Some(gtao_pipeline),
            Some(camera_size),
        ) = (

Reply:
I don't see why not, yeah. Not sure why I didn't to begin with.